transaction-impl/src/main/java/com/example/auction/transaction/impl/TransactionRepository.java

```diff
@@ -0,0 +1,195 @@
+                                "WHERE (creatorId = ? OR winnerId = ?) AND status = ? " +
+                                "ORDER BY itemId DESC " +
+                                "LIMIT ?",
+                        userId,
+                        userId,
+                        status,
+                        limit
+                )
+                .thenApply(List::stream)
+                .thenApply(rows -> rows.skip(offset))
+                .thenApply(rows -> rows.map(TransactionRepository::toTransactionSummary))
+                .thenApply(transactionSummaries -> transactionSummaries.collect(Collectors.toList()))
+                .thenApply(TreePVector::from);
+    }
+
+    private static TransactionSummary toTransactionSummary(Row transaction) {
+        return new TransactionSummary(
+                transaction.getUUID("itemId"),
+                transaction.getUUID("creatorId"),
+                transaction.getUUID("winnerId"),
+                transaction.getString("itemTitle"),
+                transaction.getString("currencyId"),
+                transaction.getInt("itemPrice"),
+                transaction.get("status", TransactionInfoStatus.class)
+        );
+    }
+
+    private static class TransactionEventProcessor extends ReadSideProcessor<TransactionEvent> {
+        private final CassandraSession session;
+        private final CassandraReadSide readSide;
+
+        private PreparedStatement insertTransactionStatement;
+
+        @Inject
+        public TransactionEventProcessor(CassandraSession session, CassandraReadSide readSide) {
+            this.session = session;
+            this.readSide = readSide;
+        }
+
+        @Override
+        public ReadSideHandler<TransactionEvent> buildHandler() {
+            return readSide.<TransactionEvent>builder("transactionEventOffset")
+                    .setGlobalPrepare(this::createTable)
+                    .setPrepare(tag -> prepareStatements())
+                    .setEventHandler(TransactionEvent.TransactionStarted.class,
+                            e -> insertTransaction(e.getItemId(), e.getTransaction()))
+                    .build();
+        }
+
+        @Override
+        public PSequence<AggregateEventTag<TransactionEvent>> aggregateTags() {
+            return TransactionEvent.TAG.allTags();
+        }
+
+        private CompletionStage<Done> createTable() {
+            return session.executeCreateTable(
+                    "CREATE TABLE IF NOT EXISTS transactionSummary (" +
+                            "itemId timeuuid PRIMARY KEY, " +
+                            "creatorId UUID, " +
+                            "winnerId UUID, " +
+                            "itemTitle text, " +
+                            "currencyId text, " +
+                            "itemPrice int, " +
+                            "status text" +
+                            ") " +
+                            "WITH CLUSTERING ORDER BY (itemId DESC)"
```

Comment:
As I mentioned on the call, Cassandra is much more strict than most databases about what fields you can use in the `WHERE` clause of a query.  In particular, the primary key needs to match the key you are using to filter your query. So in this case, we do not want the `itemId` to be the primary key, since we are not querying by `itemId`.

Cassandra does not even support `OR` in queries, so querying by `creatorId = ? OR winnerId = ?` unfortunately is not allowed.

Instead, what you can do is create a table called `userTransactions` with a `userId` column as the ["partition key"](http://docs.datastax.com/en/glossary/doc/glossary/gloss_partition_key.html).

Something like this:

```cql
CREATE TABLE IF NOT EXISTS userTransactions (
  userId UUID,
  itemId timeuuid,
  creatorId UUID,
  winnerId UUID,
  itemTitle text,
  currencyId text,
  itemPrice int,
  status text,
  PRIMARY KEY (userId, status, itemId)
)
WITH CLUSTERING ORDER BY (status ASC, itemId DESC)
```

This sets `PRIMARY KEY (userId, status, itemId)` because we need to filter by `userId` and `status`, and order by `status` and `itemId` (`ORDER BY status` is required due to https://issues.apache.org/jira/browse/CASSANDRA-10271)

When processing `TransactionStarted` events, you'll need to insert twice: once with `userId` set to `creatorId` and once with `userId` set to `winnerId`. This lets you avoid the unsupported `OR` clause in the query.

So, creating and querying tables in Cassandra is more complex than PostgreSQL or MySQL, but for that price you get very good performance and scalability. It takes a while and some trial and error to get used to it though. I'm still not quite used to it myself! :smile:

I would recommend downloading [DataStax DevCenter](https://docs.datastax.com/en/developer/devcenter/doc/devcenter/dcInstallation.html) or [DBeaver Enterprise Edition](http://dbeaver.jkiss.org/download/enterprise/). These are GUI tools that let you explore your Cassandra database. You can test creating tables, inserting data into them, and querying them from one of these tools, without the overhead of having to make changes to the Java code each time you want to try something new.

Also, as I said, adding tests for the event processor can help. The tests will let you feed data to the processor without having to finish implementing transactions in the web UI. You can look at [`ItemEventProcessorTest`](https://github.com/lagom/online-auction-java/blob/master/item-impl/src/test/java/com/example/auction/item/impl/ItemEventProcessorTest.java) (which should be renamed to `ItemRepositoryTest`) as a guide.

If you get stuck again, that's OK. Just push what you have and let me know, and I can help out more.